inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager.py

```diff
@@ -66,23 +70,36 @@ def remove(self, parameters: SourceParameters) -> None:
                         logger.debug(f"Removed source: {line}")
 
         except OSError as e:
-            # Wrap any OSError exceptions in a DispatcherException and re-raise.
-            logger.error(f"Error occurred while trying to remove sources: {e}")
-            raise DispatcherException(f"Error occurred while trying to remove sources: {e}") from e
+            raise SourceError(f"Error occurred while trying to remove sources: {e}") from e
 
     def update(self, parameters: SourceParameters) -> None:
         """Updates a source in the Ubuntu OS source file /etc/apt/sources.list"""
         # TODO: Add functionality to update a source in Ubuntu file under /etc/apt/sources.list file
-        logger.debug(f"sources: {parameters.sources}")
+        try:
+            with open(UBUNTU_APT_SOURCES_LIST, "r") as file:
+                lines = file.readlines()
+
+            sources_to_update = set(source.strip() for source in parameters.sources)
+
+        except OSError as e:
+            raise SourceError(f"Error occurred while trying to update sources: {e}") from e
 
 
 class UbuntuApplicationSourceManager(ApplicationSourceManager):
     def __init__(self) -> None:
         pass
 
     def add(self, parameters: ApplicationAddSourceParameters) -> None:
-        """Adds new application source along with its key"""
-        pass
+        # Step 1: Add key
+        key_id = add_gpg_key(parameters.gpg_key_path, parameters.gpg_key_name)
+
+        # Step 2: Add the source
+        try:
+            create_file_with_contents(os.path.join(UBUNTU_APT_SOURCES_LIST_D, parameters.file_name),
+                                      parameters.sources)
+        except (IOError, OSError) as e:
+            remove_gpg_key(key_id)
+            raise SourceError(f"Error adding application source list: {e}")
```

Comment:
(minor) could be `raise SourceError(f"Error adding application source list: {e}") from e`